agent/src/util/device_plugin_service.rs

```diff
@@ -403,100 +809,337 @@ impl InstanceDevicePlugin {
+    // a virtual device is available if it's Free or ReservedByConfiguration with requested vdev_ids
+    let mut free_device_usage_states = HashMap::new();
+    let mut reserved_by_configuration_usage_states = HashMap::new();
+    for (instance_name, device_usage_state) in device_usage_states {
+        for (id, state) in device_usage_state {
+            match state {
+                DeviceUsageStatus::Free => {
+                    free_device_usage_states
+                        .entry(instance_name.to_string())
+                        .or_insert(HashSet::new())
+                        .insert(id);
+                }
+                DeviceUsageStatus::ReservedByConfiguration(vdev_id) => {
+                    if requested_vdev_ids.contains(&vdev_id) {
+                        reserved_by_configuration_usage_states
+                            .entry(instance_name.to_string())
+                            .or_insert(HashMap::new())
+                            .insert(id, vdev_id);
+                    }
+                }
+
+                _ => (),
+            };
+        }
+    }
+
+    // Find (Instance, usage slot) for vdev ids, (Instance, usage slot) in free_device_usage_states and
+    // reserved_by_configuration_usage_states are available to be assigned.
+    // Iterate vdev ids to look up vdev id from reserved_by_configuration_usage_states, if not found
+    // pick one (Instance, usage slot) from free_device_usage_states
+    let pending_device_ids = requested_vdev_ids
+        .into_iter()
+        .filter(|vdev_id| {
+            // true means pending
+            let mut resource: Option<(String, String)> = None;
+
+            for (inst, slots) in &reserved_by_configuration_usage_states {
+                for (slot, vid) in slots {
+                    if vdev_id == vid {
+                        resource = Some((inst.clone(), slot.clone()));
+                        break;
+                    }
+                }
+                if resource.is_some() {
+                    break;
+                }
+            }
+
+            if resource.is_none() {
+                if let Some((inst, slots)) = &free_device_usage_states
+                    .iter()
+                    .max_by(|a, b| a.1.len().cmp(&b.1.len()))
+                {
+                    if let Some(slot) = slots.iter().next() {
+                        resource = Some((inst.to_string(), slot.to_string()));
+                    }
+                }
+            }
+
+            match resource {
+                Some((inst, slot)) => {
+                    reserved_by_configuration_usage_states.remove(&inst);
+                    free_device_usage_states.remove(&inst);
+                    usage_ids_to_use.insert(vdev_id.clone(), (inst, slot));
+                    false
+                }
+                None => true,
+            }
+        })
+        .collect::<Vec<String>>();
+
+    // Return error if any pending requested vdev id
```

Comment:
i find 'pending' to be a little confusing here.  i think, in this context, pending means that the device plugin failed to allocate because there weren't enough free slots?  if so, maybe pending can be renamed to unallocated or unallocatable or something like that?

Reply:
rename pending_device_ids to unallocated_device_ids